File: src/cxx/VulkanLib/Device/PhysicalDevice/PhysicalDevice.cpp
//
// Created by kgaft on 3/11/24.
//
#include "PhysicalDevice.hpp"

std::shared_ptr<std::vector<std::shared_ptr<PhysicalDevice>>> PhysicalDevice::getDevices(Instance &instance) {
    std::shared_ptr<std::vector<std::shared_ptr<PhysicalDevice>>> physicalDevices(
            new std::vector<std::shared_ptr<PhysicalDevice>>);
    uint32_t physDevCount;
    vk::Result res = instance.getInstance().enumeratePhysicalDevices(&physDevCount, nullptr);
    if(res!=vk::Result::eSuccess){
        throw std::runtime_error("Failed to find any devices");
    }
    physicalDevices->resize(physDevCount);
    std::vector<vk::PhysicalDevice> tempDevices;
    tempDevices.resize(physDevCount);
    res = instance.getInstance().enumeratePhysicalDevices(&physDevCount,
                                                          tempDevices.data());
    for (uint32_t i = 0; i < tempDevices.size(); ++i) {
        (*physicalDevices)[i] = std::shared_ptr<PhysicalDevice>(new PhysicalDevice(tempDevices[i]));
    }
    return physicalDevices;

}

PhysicalDevice::PhysicalDevice(vk::PhysicalDevice base) : base(base) {
    uint32_t propertyCount;
    vk::Result res;
    res = base.enumerateDeviceExtensionProperties(nullptr, &propertyCount, nullptr);
    if(res!=vk::Result::eSuccess){
        throw std::runtime_error("Failed to find any extensions");
    }
    extensionProperties.resize(propertyCount);
    res = base.enumerateDeviceExtensionProperties(nullptr, &propertyCount, extensionProperties.data());
    if(res!=vk::Result::eSuccess){
        throw std::runtime_error("Failed to find any extensions");
    }
    base.getFeatures(&features);
    base.getProperties(&properties);
    uint32_t queueFamilyCount;
    base.getQueueFamilyProperties(&queueFamilyCount, nullptr);
    queueProperties.resize(queueFamilyCount);
    base.getQueueFamilyProperties(&queueFamilyCount, queueProperties.data());
    vk::PhysicalDeviceProperties2 prop2{};
    prop2.pNext = &rayTracingPipelinePropertiesKhr;
    base.getProperties2(&prop2);

}

void PhysicalDevice::init(vk::PhysicalDevice base) {
    vk::Result res;
    PhysicalDevice::base = base;
    uint32_t propertyCount;
    res = base.enumerateDeviceExtensionProperties(nullptr, &propertyCount, nullptr);
    if(res!=vk::Result::eSuccess){
        std::cerr<<"Failed to gather some device info"<<std::endl;
    }
    extensionProperties.resize(propertyCount);
    res = base.enumerateDeviceExtensionProperties(nullptr, &propertyCount, extensionProperties.data());
    if(res!=vk::Result::eSuccess){
        std::cerr<<"Failed to gather some device info"<<std::endl;
    }
    base.getFeatures(&features);
    base.getProperties(&properties);
    vk::PhysicalDeviceProperties2 prop2{};
    prop2.pNext = &rayTracingPipelinePropertiesKhr;
    base.getProperties2(&prop2);

}

vk::PhysicalDevice &PhysicalDevice::getBase() {
    return base;
}

const vk::PhysicalDeviceProperties &PhysicalDevice::getProperties() const {
    return properties;
}

const vk::PhysicalDeviceFeatures &PhysicalDevice::getFeatures() const {
    return features;
}

const std::vector<vk::ExtensionProperties> &PhysicalDevice::getExtensionProperties() const {
    return extensionProperties;
}

const vk::PhysicalDeviceRayTracingPipelinePropertiesKHR &PhysicalDevice::getRayTracingPipelinePropertiesKhr() const {
    return rayTracingPipelinePropertiesKhr;
}

File: src/main/java/com/kgaft/VulkanLib/Device/PhysicalDevice/PhysicalDevice.java
package com.kgaft.VulkanLib.Device.PhysicalDevice;

import com.kgaft.VulkanLib.Instance.Instance;
import com.kgaft.VulkanLib.Utils.LwjglObject;
import com.kgaft.VulkanLib.Utils.VkErrorException;
import org.lwjgl.PointerBuffer;
import org.lwjgl.vulkan.*;

import java.lang.instrument.IllegalClassFormatException;
import java.util.ArrayList;
import java.util.List;

import static org.lwjgl.vulkan.VK13.*;

public class PhysicalDevice {
    private static List<PhysicalDevice> physicalDevices = new ArrayList<>();

    public static List<PhysicalDevice> getPhysicalDevices(Instance instance) throws VkErrorException {
        if (physicalDevices.isEmpty()) {
            int[] physicalDevicesCount = new int[1];
            vkEnumeratePhysicalDevices(instance.getInstance(), physicalDevicesCount, null);
            PointerBuffer pb = PointerBuffer.allocateDirect(physicalDevicesCount[0]);
            vkEnumeratePhysicalDevices(instance.getInstance(), physicalDevicesCount, pb);
            while (pb.hasRemaining()) {
                VkPhysicalDevice device = new VkPhysicalDevice(pb.get(), instance.getInstance());
                physicalDevices.add(new PhysicalDevice(device));
            }
        }
        return physicalDevices;
    }

    private VkPhysicalDevice base;
    private LwjglObject<VkPhysicalDeviceProperties> properties;
    private LwjglObject<VkPhysicalDeviceFeatures> features;
    private LwjglObject<VkExtensionProperties.Buffer> extensionProperties;
    private LwjglObject<VkQueueFamilyProperties.Buffer> queueProperties;
    private LwjglObject<VkPhysicalDeviceRayTracingPipelinePropertiesKHR> rayTracingPipelineProperties;

    public PhysicalDevice(VkPhysicalDevice base) throws VkErrorException {
        this.base = base;
        init(base);
    }

    public PhysicalDevice() {

    }

    public void init(VkPhysicalDevice device) throws  VkErrorException {
        try{
            int[] propertyCount = new int[1];
            int res;
            res = vkEnumerateDeviceExtensionProperties(device, (CharSequence) null, propertyCount, null);

            if (res != VK_SUCCESS) {
                throw new VkErrorException("Failed to find any extensions ", res);
            }
            extensionProperties = new LwjglObject<>(VkExtensionProperties.class, VkExtensionProperties.Buffer.class, propertyCount[0]);
            res = vkEnumerateDeviceExtensionProperties(device, (CharSequence) null, propertyCount, extensionProperties.get());
            if (res != VK_SUCCESS) {
                throw new VkErrorException("Failed to find any extensions ", res);
            }
            features = new LwjglObject<>(VkPhysicalDeviceFeatures.class);
            vkGetPhysicalDeviceFeatures(device, features.get());
            properties = new LwjglObject<>(VkPhysicalDeviceProperties.class);
            vkGetPhysicalDeviceProperties(device, properties.get());
            int[] queueFamilyCount = new int[1];
            vkGetPhysicalDeviceQueueFamilyProperties(device, queueFamilyCount, null);
            queueProperties = new LwjglObject<>(VkQueueFamilyProperties.class, VkQueueFamilyProperties.Buffer.class, queueFamilyCount[0]);
            vkGetPhysicalDeviceQueueFamilyProperties(device, queueFamilyCount, queueProperties.get());
            LwjglObject<VkPhysicalDeviceProperties2> prop2 = new LwjglObject<>(VkPhysicalDeviceProperties2.class);
            rayTracingPipelineProperties = new LwjglObject<>(VkPhysicalDeviceRayTracingPipelinePropertiesKHR.class);
            prop2.get().pNext(rayTracingPipelineProperties.get());
            prop2.get().sType$Default();
            rayTracingPipelineProperties.get().sType$Default();
            vkGetPhysicalDeviceProperties2(device, prop2.get());
        }catch (Exception e){
            e.printStackTrace();
        }

    }

    public VkPhysicalDevice getBase() {
        return base;
    }

    public LwjglObject<VkPhysicalDeviceProperties> getProperties() {
        return properties;
    }

    public LwjglObject<VkPhysicalDeviceFeatures> getFeatures() {
        return features;
    }

    public LwjglObject<VkExtensionProperties.Buffer> getExtensionProperties() {
        return extensionProperties;
    }

    public LwjglObject<VkQueueFamilyProperties.Buffer> getQueueProperties() {
        return queueProperties;
    }

    public LwjglObject<VkPhysicalDeviceRayTracingPipelinePropertiesKHR> getRayTracingPipelineProperties() {
        return rayTracingPipelineProperties;
    }

}
